app.py

```diff
@@ -12,17 +12,17 @@ def validate(keyboard, pin):
     :return: nothing or raises ValueError if validation fails
     """
     if not all([isinstance(keyboard, list), isinstance(pin, str)]):
-        raise ValueError(
+        raise TypeError(
             f"Input is not valid. Keyboard must be a list. Pin must be a string."
         )
-    if not pin.isdigit():
-        raise ValueError(f"Input is not valid. Pin must be a string of digits.")
-    if not all([isinstance(el, list) for el in keyboard]) or not all(
-        [isinstance(el[i], str) for el in keyboard for i in range(len(el))]
-    ):
-        raise ValueError(
-            f"Input is not valid. Keyboard must be a list of lists of strings."
-        )
+    if keyboard == [] or pin == "":
+        raise ValueError("Keyboard or PIN cannot be empty.")
+    if not all([isinstance(row, list) for row in keyboard]):
+        raise TypeError("Keyboard must be a list of lists.")
+    if not all([isinstance(key, str) for row in keyboard for key in row]):
+        raise TypeError("Keyboard must be a list of lists of strings.")
+    if len(set(pin) - set(key for lst in keyboard for key in lst)) != 0:
```

Comment:
Checking for length of set is atypical, most likely you want to know if it's empty or not and an empty set is a "falsy" value. Check this stackoverflow answer: https://stackoverflow.com/a/39984051/7373966

Reply:
Thanks, acknowledged!